benchmarks/README.md

```diff
@@ -0,0 +1,15 @@
+# Benchmarks, to be run by codspeed.io in CI
+
+## Running benchmarks
+
+Note that as of mid-2024, codspeed.io uses a simulated CPU (Cachegrind) to run tests, so the measures of performance are not suitable for optimizing low-level compiled code.
+
+In CI, we use `tox -e benchmarks-ci` to run the benchmarks.
+
+You can run benchmarks locally by installing `pytest-benchmark` and then running `pytest benchmarks/`.
```

Comment:
I know that this info is on codspeed ,
but maybe it's useful to have it here

```suggestion
You can run benchmarks locally by installing `pytest-benchmark` and then running `pytest benchmarks/`.

No reports are generated.
This is just to check the tests aren't failing.
```

Reply:
Would it make sense to generate relative comparisons by just running `time` or something? Seems like an unfortunate detail that there's no information at all provided in local development.

Reply:
My understanding is that it's hard to create any meaningful comparison.

You can have a test-suite execute on trunk and then on this branch, and comparate the time... but I don't know if the error margin is good enough.

--------

so the remaining option is to run the test with valgrid in trunk and on the current branch... that this requires a lot of work... and I would say that such work is outside of the scope of this PR.

this PR is about getting some reports via codspeed.io

-----------

I would say that we should give codspeed.io a try and see if it help.

Maybe, we will find out that this is not very helpful and we need to implement our own tools to run and colect results from `valgrind --tool=cachegrind python -m twisted.trial`

Reply:
> I would be nice to have some local results... but.
> 
> My understanding is that it's hard to create any meaningful comparison.
> 
> You can have a test-suite execute on trunk and then on this branch, and comparate the time... but I don't know if the error margin is good enough.

"good enough" for what?  If I'm developing a performance improvement, I don't want to push to CI and wait for 20 minutes to see if my changes made any impact at all.  If I'm running repeatedly in the exact same environment on my development workstation, the results are same enough to get a comparison.  I am not even talking about running on trunk, then a branch; I just mean being able to run a benchmark, make an edit, then run it again to see if the edit did anything.

Reply:
> so the remaining option is to run the test with valgrid in trunk and on the current branch... that this requires a lot of work... and I would say that such work is outside of the scope of this PR.

I'm totally fine with the precision and rigor of valgrind being reserved for running in CI.

Reply:
> I would say that we should give codspeed.io a try and see if it help.

Sure, I just don't want to structure our benchmarks so they *can't* be run locally, given how simple our requirements are at this point. It's a bit annoying to have to drag in all of pytest but I'm not going to complain if that's the API that a service that someone *else* maintains presents to us

Reply:
To be clear I would not block this on this basis, I think "run function in a loop, print how long it took" should *not* be an unreasonable demand but there seem to be a lot of moving pieces with codspeed so if it is actually difficult then go ahead and land without it

Reply:
`pytest-benchmark` does actually output numbers, so if you use it instead of `pytest-codspeed` you're good for local development.